### PR TITLE
fix(auth): allow removing non-last admin users

### PR DESCRIPTION
**Pull request by vishu-bh**
## 🔗 Related Issue
Closes # https://github.com/IBM/mcp-context-forge/issues/2808

---

## 📝 Summary
_What does this PR do and why?_

This PR fixes thes issue When an admin user tries to remove the admin privileges from the other admin user, API returns error message. It should not return an error message untill it is the last admin.

Key Changes:
- Updated the admin permission removal logic, to check first if it is the last admin. If it is last admin then admin permission cannot be removed.
- if the requesting admin is trying to self demote itself using api or UI, then an error message will be shown
- Lastly, if admin tries to demote other admin and it is not the last admin in system then it should be allowed to do it. 

The root cause was: There is a flag in config called 'protect_all_admins'. This flag is set to true by in configuration. Hence, it was restricting other admins to demote other admins. 

---

## 🏷️ Type of Change
- [x] Bug fix
- [ ] Feature / Enhancement
- [ ] Documentation
- [ ] Refactor
- [ ] Chore (deps, CI, tooling)
- [ ] Other (describe below)

---

## 🧪 Verification

| Check                     | Command         | Status |
|---------------------------|-----------------|--------|
| Lint suite                | `make lint`     |  ✅ Passed (Pylint score: 10/10)      |
| Unit tests                | `make test`     | ✅ Passed (updated unit tests with the changes)        |
| Coverage ≥ 80%            | `make coverage` | ✅ Passed (100% for modified files, 99%+ overal        |

---

## ✅ Checklist
- [x] Code formatted (`make black isort pre-commit`)
- [x] Tests added/updated for changes
- [x] Documentation updated (if applicable)
- [x] No secrets or credentials committed

---

## 📓 Notes (optional)
_Screenshots, design decisions, or additional context._

### CODE REVIEW

**Comment by gcgoncalves** [comment]
A request from an admin user to edit it's own profile and remove the admin status is successful - all attributes get updates, except for `email` and `is_admin`. Should we raise and return an error `403 - Not Authorised` or keep the current behaviour?

**Comment by crivetimihai** [comment]
Thanks for this @vishu-bh. This addresses a valid auth issue (#2808). However, this PR currently has **merge conflicts** and CI hasn't run. Please rebase onto `main` to resolve the conflicts and trigger CI. Once the conflicts are resolved and CI is green, we can review.

**Comment by vishu-bh** [comment]
> A request from an admin user to edit it's own profile and remove the admin status is successful - all attributes get updates, except for `email` and `is_admin`. Should we raise and return an error `403 - Not Authorised` or keep the current behaviour?

As discussed we are keeping the current behavior. As admin when try to edit its own profile, cannot remove admin access as it does not see the checkbox. and if request is edited check in admin.py at at line 7329 will preserve the access even if it is removed in request. Hope that makes sense.

**Comment by vishu-bh** [comment]
> ## Code Review Summary for PR #3247
> **Issue**: [#2808](https://github.com/IBM/mcp-context-forge/issues/2808) - Admin user should be able to remove admin privileges from other admin users
> 
> ### Overview
> This PR changes the admin protection behavior from blocking ALL admin demotions (when `protect_all_admins=True`) to specifically preventing **self-demotion** while allowing admins to demote other admins.
> 
> ### Key Changes
> 1. **Core Logic** ([`email_auth_service.py:1559`](mcp-context-forge/mcpgateway/services/email_auth_service.py:1559)): Added `requesting_user_email` parameter and self-demotion check
> 2. **API Integration** ([`email_auth.py:786`](mcp-context-forge/mcpgateway/routers/email_auth.py:786)): Passes requesting user's email to service
> 3. **UI Integration** ([`admin.py:7346`](mcp-context-forge/mcpgateway/admin.py:7346)): Passes requesting user's email from UI
> 4. **Test Updates**: Modified existing tests and added new test for successful cross-admin demotion
> 
> ### Review Findings
> **8 issues identified** (1 high, 3 medium, 4 low) - all added to Bob Findings panel:
> 
> **Critical Issues:**
> 
> * **High**: Missing test coverage for cross-admin demotion scenario
> * **Medium**: Outdated test still references old error message
> * **Medium**: Email normalization edge case could bypass self-demotion check
> * **Medium**: Case-insensitive email comparison may fail with empty string
> 
> **Documentation & Maintenance:**
> 
> * Configuration documentation needs updating to reflect new behavior
> * Test naming inconsistencies and misleading docstrings
> * Test fixture changes may affect other tests
> 
> ### Alignment with Issue #2808
> ✅ **Correctly addresses the issue**: The PR successfully allows admins to demote other admins while preventing self-demotion, which aligns with the expected behavior described in the issue.
> 
> ### Recommendations
> 1. Add explicit None check for `requesting_user_email` before string operations
> 2. Update [`config.py:511`](mcp-context-forge/mcpgateway/config.py:511) documentation to reflect new behavior
> 3. Fix outdated test at [`test_email_auth_router.py:809`](mcp-context-forge/tests/unit/mcpgateway/routers/test_email_auth_router.py:809)
> 4. Add test case for email normalization edge cases
> 5. Add explicit test for Admin A demoting Admin B scenario

Thanks @msureshkumar88 for the review addressed the comments

**Review by msureshkumar88 — APPROVED** [review]
## ✅ LGTM 


**Summary**: Successfully prevents admin self-demotion while allowing cross-admin operations.

**Ready to merge** 🚀

**Comment by gcgoncalves** [comment]
Reading the issue and the fix, I am not sure this is the correct solution. If the option to "protect all admins" means that only the last one should be protected, it should be renamed. Otherwise, it can be stripped altogether, and documented accordingly, otherwise it might break the user's expectations when setting it to true.

**Comment by vishu-bh** [comment]
> Reading the issue and the fix, I am not sure this is the correct solution. If the option to "protect all admins" means that only the last one should be protected, it should be renamed. Otherwise, it can be stripped altogether, and documented accordingly, otherwise it might break the user's expectations when setting it to true.

@gcgoncalves  I think idea is to no allow admin demote itself. So currently in system, admin should be able to demote other admins. And it should not be able to demote itself. protect_all_admins flag was restricting admin to admin demotion. so that check is removed and new check is added to prevent admin self demotion. 

In admin UI, if admin edit its own user, it will not see admin flag at all. But it can still do it from api. To keep the behaviour consistent check in email_service is added. So now admin cannot self demote it either from API and from admin ui. But admin can demote other admins. 

Protect_all_admin flag is not removed cause it is used in other places in the application. Let me know if it makes sense other wise we can have a quick catchup.

**Review by msureshkumar88 — DISMISSED** [review]

**Comment by gcgoncalves** [comment]
> > Reading the issue and the fix, I am not sure this is the correct solution. If the option to "protect all admins" means that only the last one should be protected, it should be renamed. Otherwise, it can be stripped altogether, and documented accordingly, otherwise it might break the user's expectations when setting it to true.
> 
> @gcgoncalves I think idea is to no allow admin demote itself. So currently in system, admin should be able to demote other admins. And it should not be able to demote itself. protect_all_admins flag was restricting admin to admin demotion. so that check is removed and new check is added to prevent admin self demotion.
> 
> In admin UI, if admin edit its own user, it will not see admin flag at all. But it can still do it from api. To keep the behaviour consistent check in email_service is added. So now admin cannot self demote it either from API and from admin ui. But admin can demote other admins.
> 
> Protect_all_admin flag is not removed cause it is used in other places in the application. Let me know if it makes sense other wise we can have a quick catchup.

The intent makes sense, but `protect_all_admins` is still a setting users can enable, and the name implies it protects all admins from demotion. After this change, `protect_all_admins=true` no longer prevents cross-admin demotion via update_user, even though users might expect it to. If the flag is used elsewhere but its semantics have changed, that should be documented, either in the config docstring or release notes.

File: mcpgateway/services/email_auth_service.py [hunk 1]
```diff
@@ -1555,8 +1557,14 @@ async def update_user(
...
                 would_lose_admin = (is_admin is not None and not is_admin) or (is_active is not None and not is_active)
                 if would_lose_admin:
-                    if settings.protect_all_admins:
-                        raise ValueError("Admin protection is enabled — cannot demote or deactivate any admin user")
+                    # Check if this is self-demotion
+                    is_self_demotion = requesting_user_email and requesting_user_email.lower().strip() == email.lower().strip()
+
+                    # Block self-demotion
+                    if is_self_demotion:
+                        raise ValueError("Cannot demote or deactivate your own admin account")
```
Comment (gcgoncalves) [inline mcpgateway/services/email_auth_service.py:1693]: When `requesting_user_email=None` (the default), `is_self_demotion` is set to `None` and the entire this block is skipped. A caller can demote an admin without restriction by not passing a `requesting_user_email`. The parameter should either be made required (non-optional) or the logic should default to blocking: `is_self_demotion = not requesting_user_email` or `email_a == email_b`.

File: tests/unit/mcpgateway/services/test_email_auth_basic.py [hunk 2]
```diff
@@ -2315,6 +2387,83 @@ async def test_update_user_blocks_demote_last_active_admin(self, service, mock_d
...
+                    role_id="admin_role_id",
+                    scope="global",
+                    scope_id=None
+                )
+                mock_role_service_instance.assign_role_to_user.assert_called_once_with(
+                    user_email="admin-b@example.com",
+                    role_id="user_role_id",
+                    scope="global",
+                    scope_id=None,
+                    granted_by="admin-b@example.com"
```
Comment (gcgoncalves) [inline tests/unit/mcpgateway/services/test_email_auth_basic.py]: `granted_by` is the email of the user being demoted, not the admin performing the action ("admin-a@example.com"). The same issue appears in `test_update_user_demote_admin_success`, where the requester is `superadmin@example.com` but `granted_by="admin@example.com"` (the target).

If these assertions reflect what the service actually does, the audit trail records the role change as self-granted by the target user rather than by the acting admin. Could you confirm what value is passed to `granted_by` in the service code? If it's pulling from the target user's email instead of `requesting_user_email`, that's an audit integrity bug worth fixing here.

Comment (vishu-bh) [reply tests/unit/mcpgateway/services/test_email_auth_basic.py]: adrressed

File: mcpgateway/services/email_auth_service.py (hunk 1 above)
Comment (vishu-bh) [reply mcpgateway/services/email_auth_service.py:1693]: added a check to raise error if no requesting_user_email found

**Review by msureshkumar88 — DISMISSED** [review]

**Comment by jonpspri** [comment]
Based on repeated review, I'm rethinking this too.  I think we'll close this PR and send 2808 back to the drawing board.
